bodhi/server/consumers/greenwave.py

```diff
@@ -0,0 +1,66 @@
+# Copyright © 2019 Red Hat, Inc.
+#
+# This file is part of Bodhi.
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along with
+# this program; if not, write to the Free Software Foundation, Inc., 51
+# Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+"""
+The "greenwave handler".
+
+This module is responsible for listening for messages from greenwave.
+It then updates the policies of the build that greenwave checked.
+"""
+
+import logging
+
+import fedora_messaging
+
+from bodhi.server.models import Build
+from bodhi.server.util import transactional_session_maker
+from bodhi.server.exceptions import BodhiException
+
+log = logging.getLogger(__name__)
+
+
+class GreenwaveHandler:
+    """
+    The Bodhi Greenwave Handler.
+
+    A fedora-messaging listener waiting for messages from greenwave about enforced policies.
+    """
+
+    def __init__(self):
+        """Initialize the GreenwaveHandler."""
+        self.db_factory = transactional_session_maker()
+
+    def __call__(self, message: fedora_messaging.api.Message):
+        """Handle messages arriving with the configured topic."""
+        msg = message.body.get("msg")
+        if msg is None:
+            log.debug("Ignoring message without 'msg'.")
+            return
+
+        subject_identifier = msg.get("subject_identifier")
+
+        if subject_identifier is None:
+            raise BodhiException("Couldn't find subject_identifier in the Greenwave message")
+
+        with self.db_factory():
+
+            build = Build.get(subject_identifier)
+            if build is None:
+                raise BodhiException(f"Couldn't find build {subject_identifier} in DB")
+
+            log.info(f"Updating the test_gating_status for: {build.update.alias}")
+            build.update.update_test_gating_status()
```

Comment:
I think this needs to be committed to the database, e.g. like this:

```
    with self.db_factory() as dbsession:
        ...
        build.update.update_test_gating_status()
        dbsession.commit()
```

@cverna?

Reply:
yes sounds good to me :+1:

Reply:
+1 from me too.
Always good to specifically push/commit/close the object and get everything written before you move onto the next one.  Good catch.

Reply:
I'm getting all sorts of `"name 'dbsession' is not defined"`  and `"... object has no attribute 'dbsession' "`
I've tried lots of variations on dbsession (db.session, db_session, db, session)
I've tried having it be from things (self.dbsession, build.dbsession, self.db_factory.dbsession).
And a very good spread of the combination of the above two things.

Reply:
@tdawson it is because you need to change the `with self.db_factory():` statement to `with self.db_factory() as dbsession:` then you can use `dbsession` :+1:

Reply:
That works.  Thanks

Reply:
Although I did add the dbsession.commit(), and things worked, after reading bowlofeggs comment below, it looks like it is just adding a duplicate call.  So I have removed it.

Reply:
> Although I did add the dbsession.commit(), and things worked, after reading bowlofeggs comment below, it looks like it is just adding a duplicate call. So I have removed it.

Ha nice did not know that :-)